docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,141 @@
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/logging.edl" import *;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+In example (1), linking would fail if the enclave made a call to `oe_realloc_ocall()`
+(also part of tee.edl) unless the developer modified the import line to
+`from "openenclave/edl/tee.edl" import oe_log_ocall, oe_realloc_ocall;`. In this scenario,
+example (2) will work without modification.
+
+## EDL local structures
+
+Under this design, system EDL files will prefer using local structures; that is, structures
+that are defined in EDL. This is distinct from foreign structures, which are defined in
+C/C++ header files and included in the EDL file.
+
+In the case that a foreign structure must be used, system EDL files may only include headers
+that are installed with the SDK.
+
+Local EDL structures provide additional security benefits detailed in the
+[Full EDL Serialzation design doc](
+https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/full_edl_serialization.md#motivation)
+so foreign structures should be avoided whenever possible.
+
+## Developer opt-out
+
+Because system OCalls are generally called by the SDK implementation, developers
+may not have direct control over which OCalls will be called by their application.
+A developer can refuse to import a function in EDL which will cause a linker error
+if it is used, but they should also be able to take action to resolve this failure.
+
+To this end, any OE SDK feature which makes an OCall on behalf of an enclave
+should provide a mechanism to disable the behavior which makes an OCall.
+
+For example, the switchless calls feature makes an OCall
+`oe_wake_switchless_worker_ocall()`. If there is code compiled into the enclave
+which calls this function, the enclave developer will be forced to import that
+function into their own EDL. To allow the developer to opt out of this OCall,
+it should be possible for the enclave developer to turn off the switchless calls
+feature such that an enclave which does not use switchless calls will not require
+this OCall. This mechanism should not require the developer to recompile the OE
+SDK.
+
+## Exceptions to EDL import and opt-in/out mechanisms
+
+A small number of OCalls in OE SDK are not implemented in EDL. Until these are
+converted to EDL, these OCalls will always be included and the developer cannot
+opt out of them.
+
+The following 4 OCalls can be implemented in EDL, they just aren't today. Issue
+`#2008` tracks converting these to EDL.
+
+* `OE_OCALL_MALLOC`
+* `OE_OCALL_FREE`
+* `OE_OCALL_THREAD_WAIT`
+* `OE_OCALL_THREAD_WAKE`
+
+## List of system OCalls
+
+All system OCalls (other than the previously mentioned exceptions) are found
+in one of the following EDL files
+
+* `common/attestation.edl`
+* `common/logging.edl`
+* `common/memory.edl`
+* `common/keys.edl`
```

Comment:
This edl only defines ECalls, should it be removed from this list (or there may be more such cases)?

Reply:
Ah you're right. I'll go through and remove any with no ocalls